pkg/cloudfabric-controller/controllerframework/controller_framework_test.go

```diff
@@ -290,7 +290,348 @@ func TestConsolidateControllerInstances_ReturnValues_MergeAndAutoExtends(t *test
 	assert.Equal(t, int64(math.MaxInt64), controllerInstanceBase.sortedControllerInstancesLocal[0].controllerKey)
 }
 
-func TestGetMaxInterval(t *testing.T) {
+func TestGenerateKey(t *testing.T) {
+	const TOTAL_SCOPE = int64(9223372036854775807)        // 1       100%
+	const HALF_SCOPE = int64(4611686018427387903)         // 1/2      50%
+	const ONE_FOURTH_SCOPE = int64(2305843009213693951)   // 1/4      25%
+	const THREE_FOURTH_SCOPE = int64(6917529027641081855) // 3/4      75%
+	const ONE_EIGHTH_SCOPE = int64(1152921504606846975)   // 1/8    12.5%
+	const THREE_EIGHTH_SCOPE = int64(3458764513820540927) // 3/8    37.5%
+	const FIVE_EIGHTH_SCOPE = int64(5764607523034234879)  // 5/8    62.5%
+	const SEVEN_EIGHTH_SCOPE = int64(8070450532247928831) // 7/8    87.5%
+	const ONE_SIXTEENTH_SCOPE = int64(576460752303423487) // 1/16   6.25%
+
+	controllerBase := new(ControllerBase)
+
+	// Start state: []
+	// End state: [100]
+	// controllerKey: math.MaxInt64 * 100%
+	// Description: Start state [] means there is no controller. End state [100] means the new controller will
+	//              controll 100% of the scope space. The total scope space is math.MaxInt64 (9223372036854775807).
+	//              In this way, we expect controller key is math.MaxInt64.
+	//              (Same expression logic for other test cases)
+	controllerKey := controllerBase.generateKey()
+	assert.Equal(t, TOTAL_SCOPE, controllerKey)
+
+	// Start state: [100]
+	// End state: [50, 50]
+	// controllerKey: math.MaxInt64 * 50% -- Scope of 100% splitted
+	controllerBase.sortedControllerInstancesLocal = []controllerInstanceLocal{
+		{
+			lowerboundKey: 0,
+			controllerKey: TOTAL_SCOPE,
+		},
+	}
+	controllerKey = controllerBase.generateKey()
+	assert.Equal(t, HALF_SCOPE, controllerKey)
+
+	// Start state: [50, 50]
+	// End state: [25, 25, 50]
+	// controllerKey: math.MaxInt64 * 25%  -- Scope of first 50% splitted
+	controllerBase.sortedControllerInstancesLocal = []controllerInstanceLocal{
```

Comment:
Generate sortedControllerInstancesLocal can be extracted to a util

Reply:
New test case added